### PR TITLE
oracle-instantclient: 12.2.0.1.0 -> 19.3.0.0.0

### PR DESCRIPTION
**Pull request by flokli**
##### Motivation for this change
It's now possible to download Oracle Instant Client without having to
click a checkbox to agree to the license, and then log in with an Oracle
account to start the download, at least on Linux (for now).

Darwin still seems to be at 18.1 (and needs manual fetching).
Version numbers haven't been that aligned recently, so I refactored the derivation to support different version numbers and fetchers per platform.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Repro:

```sh
$(nix-build -E "with import /path/to/checkout {};python3.withPackages (p: [p.cx_oracle p.ptpython])")/bin/ptpython
```
```py
import cx_Oracle
connection = cx_Oracle.connect("MAIN_USER/MAIN_PASSWORD@localhost/orclpdb1")
```

### CODE REVIEW

File: pkgs/development/libraries/oracle-instantclient/default.nix [hunk 1]
```diff
@@ -1,71 +1,130 @@
...
-    install -Dm644 sdk/demo/* $out/share/${name}/demo
+    install -Dm644 sdk/demo/* $out/share/${pname}-${version}/demo
+
+    # provide alias
+    ln -sfn $out/bin/sqlplus $out/bin/sqlplus64
 
     # PECL::oci8 will not build without this
     # this symlink only exists in dist zipfiles for some platforms
-    ln -sfn $out/lib/libclntsh${extLib}.12.1 $out/lib/libclntsh${extLib}
+    ln -sfn $lib/lib/libclntsh${extLib}.12.1 $lib/lib/libclntsh${extLib}
```
Comment (aanderse) [inline pkgs/development/libraries/oracle-instantclient/default.nix]: `12.1` would need to be updated here I imagine.

**Review by aanderse — CHANGES_REQUESTED** [review]
Will test once I get a chance.

File: pkgs/development/libraries/oracle-instantclient/default.nix (hunk 1 above)
Comment (flokli) [reply pkgs/development/libraries/oracle-instantclient/default.nix]: Not sure. The package ships multiple versions:

```
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntsh.so
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntsh.so.10.1
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntsh.so.11.1
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntsh.so.12.1
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntsh.so.18.1
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntsh.so.19.1
/nix/store/a9bwgf3w7xq1019kacjbwkfssp1g7n7y-oracle-instantclient-19.3.0.0.0-lib/lib/libclntshcore.so.19.1
```

I think the better fix would be to get PECL::oci8 to `dlopen`/link against the version it works with.

Comment (flokli) [reply pkgs/development/libraries/oracle-instantclient/default.nix]: building on current master and `readelf -d` at the `oci8.so` shows it only needs the followin libraries:

```
 0x0000000000000001 (NEEDED)             Shared library: [libclntsh.so.12.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```

so we can drop the `libclntsh.so` symlink.

However, there's another problem in `phpPackages.oci8` - it didn't handle oracle-instantclient being a split derivation. Will upload a patch fixing this.

**Comment by flokli** [comment]
Was able to build `perlPackages.DBDOracle`, `phpPackages.oci8` and `python3.pkgs.cx_oracle` successfully.

Poked at the .so files on linux, and tried connecting through the python bindings. Looks good so far :-)

@aanderse, could you take a look at situation on darwin? At least the python bindings previously did work previously…

**Review by aanderse — CHANGES_REQUESTED** [review]
After making the mentioned changes to your branch I was able to compile the following on `darwin`:
- `oracle-instantclient`
- `php72Packages.oci8`
- `php73Packages.oci8`
- `python3.pkgs.cx_oracle`

And without making any changes I was able to compile the following on `linux`:
- `oracle-instantclient`
- `php72Packages.oci8`
- `php73Packages.oci8`
- `python3.pkgs.cx_oracle`
- `perlPackages.DBDOracle`
- `perldevelPackages.DBDOracle`

`perldevelPackages.DBDOracle` had the following failure on `darwin`:
```
#   Failed test 'require DBDOracleTestLib;'
#   at t/00dbdoracletestlib.t line 11.
#     Tried to require 'DBDOracleTestLib'.
#     Error:  Can't load '/private/var/folders/3y/f246ctxd07g_ght97sqhlhs80000gp/T/nix-build-perl5.31.2-DBD-Oracle-1.80.drv-0/DBD-Oracle-1.76/blib/arch/auto/DBD/Oracle/Oracle.bundle' for module DBD::Oracle: dlopen(/private/var/folders/3y/f246ctxd07g_ght97sqhlhs80000gp/T/nix-build-perl5.31.2-DBD-Oracle-1.80.drv-0/DBD-Oracle-1.76/blib/arch/auto/DBD/Oracle/Oracle.bundle, 2): Library not loaded: @rpath/libclntsh.dylib.18.1
#   Referenced from: /private/var/folders/3y/f246ctxd07g_ght97sqhlhs80000gp/T/nix-build-perl5.31.2-DBD-Oracle-1.80.drv-0/DBD-Oracle-1.76/blib/arch/auto/DBD/Oracle/Oracle.bundle
#   Reason: image not found at /nix/store/q9dgm21xc01as964s5fj5kqyq9cs0ivs-perl-5.31.2/lib/perl5/5.31.2/darwin-2level/DynaLoader.pm line 197.
#  at t/lib/DBDOracleTestLib.pm line 29.
```

File: pkgs/development/libraries/oracle-instantclient/default.nix [hunk 2]
```diff
@@ -1,71 +1,123 @@
...
+
+  # hashes per component and architecture
+  hashes = {
+    "x86_64-linux" = {
+      "basic"   = "1yk4ng3a9ka1mzgfph9br6rwclagbgfvmg6kja11nl5dapxdzaxy";
+      "sdk"     = "115v1gqr0czy7dcf2idwxhc6ja5b0nind0mf1rn8iawgrw560l99";
+      "sqlplus" = "0zj5h84ypv4n4678kfix6jih9yakb277l9hc0819iddc0a5slbi5";
+      "odbc"    = "1g1z6pdn76dp440fh49pm8ijfgjazx4cvxdi665fsr62h62xkvch";
+    };
+    "x86_64-darwin" = {
```
[the remark this answers is not in the document]
Comment (flokli) [reply pkgs/development/libraries/oracle-instantclient/default.nix]: added, thanks.

File: pkgs/development/libraries/oracle-instantclient/default.nix [hunk 3]
```diff
@@ -1,71 +1,123 @@
...
+    url = "https://download.oracle.com/otn_software/linux/instantclient/193000/${srcFilename}";
+    sha256 = hash;
+  };
+
+  # pick the appropriate fetcher depending on the platform
+  fetcher = if stdenv.hostPlatform.system == "x86_64-linux" then fetchSimple else fetchClickThrough;
+
+  # assemble srcs
+  srcs = map (component:
+    (fetcher (srcFilename component arch version "") hashes."${component}" or ""))
```
[the remark this answers is not in the document]
Comment (flokli) [reply pkgs/development/libraries/oracle-instantclient/default.nix]: fixed, thanks.

File: pkgs/development/libraries/oracle-instantclient/default.nix [hunk 4]
```diff
@@ -1,71 +1,123 @@
...
+    # provide alias
+    ln -sfn $out/bin/sqlplus $out/bin/sqlplus64
+  '';
 
-    # PECL::oci8 will not build without this
-    # this symlink only exists in dist zipfiles for some platforms
-    ln -sfn $out/lib/libclntsh${extLib}.12.1 $out/lib/libclntsh${extLib}
+  postFixup = optionalString stdenv.isDarwin ''
+    for exe in "$out/bin/"* ; do
+      install_name_tool -add_rpath "$lib/lib" "$exe"
```
[the remark this answers is not in the document]
Comment (flokli) [reply pkgs/development/libraries/oracle-instantclient/default.nix]: right :see_no_evil: Thanks!

File: pkgs/development/libraries/oracle-instantclient/default.nix [hunk 5]
```diff
@@ -1,71 +1,123 @@
...
-  }."${stdenv.hostPlatform.system}" or throwSystem;
+  # calculate the filename of a single zip file
+  srcFilename = component: arch: version: rel:
+    "instantclient-${component}-${arch}-${version}" +
+    (optionalString (rel != "") "-${rel}") +
+    (optionalString (arch == "linux.x64") "dbru") + # ¯\_(ツ)_/¯
+    ".zip";
 
+  # fetcher for the clickthrough artifacts (requiring manual download)
+  fetchClickThrough =  srcFilename: hash: (requireFile {
```
[the remark this answers is not in the document]
Comment (flokli) [reply pkgs/development/libraries/oracle-instantclient/default.nix]: right :see_no_evil: Thanks!

**Comment by flokli** [comment]
I adressed all the changes, and could get DBDOracle to build aswell - had to add `${oracle-instantclient.lib}/lib` to the rpath of the resulting `Oracle.bundle` file. The tests now pass there, too.

**Review by aanderse — APPROVED** [review]
Successfully tested connectivity to an `oracle` database using `phpPackages.oci8`. :+1: 

Thanks @flokli!

**Pull request merged**